libgen/utils.go

```diff
@@ -0,0 +1,649 @@
+	}
+
+	spec.Linux.Seccomp = &secc
+	return nil
+}
+
+func parseArgs(args2parse string) ([]*rspec.Arg, error) {
+	var Args []*rspec.Arg
+	argstrslice := strings.Split(args2parse, ",")
+	for _, argstr := range argstrslice {
+		args := strings.Split(argstr, "/")
+		if len(args) == 4 {
+			index, err := strconv.Atoi(args[0])
+			value, err := strconv.Atoi(args[1])
+			value2, err := strconv.Atoi(args[2])
+			if err != nil {
+				return nil, err
+			}
+			switch args[3] {
+			case "":
+			case "SCMP_CMP_NE":
+			case "SCMP_CMP_LT":
+			case "SCMP_CMP_LE":
+			case "SCMP_CMP_EQ":
+			case "SCMP_CMP_GE":
+			case "SCMP_CMP_GT":
+			case "SCMP_CMP_MASKED_EQ":
+			default:
+				return nil, fmt.Errorf("seccomp-sysctl args must be empty or one of SCMP_CMP_NE|SCMP_CMP_LT|SCMP_CMP_LE|SCMP_CMP_EQ|SCMP_CMP_GE|SCMP_CMP_GT|SCMP_CMP_MASKED_EQ")
+			}
+			op := rspec.Operator(args[3])
+			Arg := rspec.Arg{
+				Index:    uint(index),
+				Value:    uint64(value),
+				ValueTwo: uint64(value2),
+				Op:       op,
+			}
+			Args = append(Args, &Arg)
+		} else {
+			return nil, fmt.Errorf("seccomp-sysctl args error: %s", argstr)
+		}
+	}
+	return Args, nil
+}
+
+// AddIDMappings add the mappings of uid and gid into spec.
+func AddIDMappings(spec *rspec.Spec, uidMaps, gidMaps []string) error {
+	for _, uidms := range uidMaps {
+		idm := strings.Split(uidms, ":")
+		if len(idm) == 3 {
+			hid, err := strconv.Atoi(idm[0])
+			cid, err := strconv.Atoi(idm[1])
+			size, err := strconv.Atoi(idm[2])
+			if err != nil {
+				return err
+			}
+			uidmapping := rspec.IDMapping{
+				HostID:      uint32(hid),
+				ContainerID: uint32(cid),
+				Size:        uint32(size),
+			}
+			spec.Linux.UIDMappings = append(spec.Linux.UIDMappings, uidmapping)
+		} else {
+			return fmt.Errorf("uidmappings error: %s", uidms)
+		}
+	}
+
+	for _, gidms := range gidMaps {
+		idm := strings.Split(gidms, ":")
+		if len(idm) == 3 {
+			hid, err := strconv.Atoi(idm[0])
+			cid, err := strconv.Atoi(idm[1])
+			size, err := strconv.Atoi(idm[2])
+			if err != nil {
+				return err
+			}
+			gidmapping := rspec.IDMapping{
+				HostID:      uint32(hid),
+				ContainerID: uint32(cid),
+				Size:        uint32(size),
+			}
+			spec.Linux.GIDMappings = append(spec.Linux.GIDMappings, gidmapping)
+		} else {
+			return fmt.Errorf("gidmappings error: %s", gidms)
+		}
+	}
+
+	return nil
+}
+
+// AddRootPropagation adds the root propagation setting into spec.
+func AddRootPropagation(spec *rspec.Spec, rp string) error {
+	switch rp {
+	case "":
+	case "private":
+	case "rprivate":
+	case "slave":
+	case "rslave":
+	case "shared":
+	case "rshared":
+	default:
+		return fmt.Errorf("rootfs-propagation must be empty or one of private|rprivate|slave|rslave|shared|rshared")
+	}
+	spec.Linux.RootfsPropagation = rp
+	return nil
+}
+
+// AddHooks add the (prestart|poststop|poststart)hooks into spec.
+func AddHooks(spec *rspec.Spec, preStartHooks, postStopHooks, postStartHooks []string) error {
```

Comment:
@wking , reconsidering the API seems a good idea.
For this PR, I was just trying to separate `generate.go` and the library, and tried to avoid drastic changes of the code.
I will change the code after the new API is decided.